src/webnn_native/nnapi/GraphNnapi.cpp

```diff
@@ -0,0 +1,644 @@
+#include "webnn_native/Utils.h"
+
+#define WEBNN_ASSERT(condition, message) \
+    do {                                 \
+        dawn::ErrorLog() << message;     \
+        DAWN_ASSERT(condition);          \
+    } while (0)
+
+namespace webnn_native { namespace nnapi {
+
+    Graph::Graph(Context* context) : GraphBase(context) {
+        operandCount = 0;
+
+        mScalarInt32Operand.type = ANEURALNETWORKS_INT32;
+        mScalarInt32Operand.dimensionCount = 0;
+        mScalarInt32Operand.dimensions = nullptr;
+        mScalarInt32Operand.scale = 0.0f;
+        mScalarInt32Operand.zeroPoint = 0;
+
+        mScalarBoolOperand.type = ANEURALNETWORKS_BOOL;
+        mScalarBoolOperand.dimensionCount = 0;
+        mScalarBoolOperand.dimensions = nullptr;
+        mScalarBoolOperand.scale = 0.0f;
+        mScalarBoolOperand.zeroPoint = 0;
+
+        mNnapiMgr = new NnapiManager();
+    }
+
+    Graph::~Graph() {
+        for (auto node : mGraphOperandInfo) {
+            if (node.second.mem)
+                mNnapiMgr->FreeMemory(node.second.mem);
+
+            close(node.second.fd);
+        }
+
+        delete mNnapiMgr;
+    }
+
+    MaybeError Graph::AddConstant(const op::Constant* constant) {
+        NodeInfo node;
+        void* buffer = const_cast<void*>(constant->GetBuffer());
+
+        CreateNodeFromOperandDescriptor(constant->GetOperandDescriptor(), &node);
+        DAWN_TRY(mNnapiMgr->CreateOperandAndSetMemory("const", &node, buffer));
+
+        mGraphOperandInfo[node.opIndex] = node;
+        mGraphNodeMap[constant->PrimaryOutput()] = node.opIndex;
+
+        return {};
+    }
+
+    MaybeError Graph::AddInput(const op::Input* input) {
+        NodeInfo node;
+
+        CreateNodeFromOperandDescriptor(input->GetOperandDescriptor(), &node);
+        DAWN_TRY(mNnapiMgr->CreateInputOutputOperand(input->GetName(), &node));
+
+        mGraphOperandInfo[node.opIndex] = node;
+        mGraphNodeMap[input->PrimaryOutput()] = node.opIndex;
+        mInputIdMap[input->GetName()] = node;
+        mGraphInputs.push_back(node.opIndex);
+
+        return {};
+    }
+
+    MaybeError Graph::AddOutput(const std::string& name, const OperandBase* output) {
+        uint32_t index = mGraphNodeMap[output];
+        auto node = mGraphOperandInfo[index];
+
+        DAWN_TRY(mNnapiMgr->CreateInputOutputOperand(name, &node, false));
+
+        mGraphOutputs.push_back(index);
+        mOutputNameMap[name] = node;
+        return {};
+    }
+
+    MaybeError Graph::AddInstanceNorm(const op::InstanceNorm* instanceNorm) {
+        DAWN_TRY(CheckStatusCode(ANEURALNETWORKS_OP_FAILED, "nnapi Instance norm"));
+        return {};
+    }
+
+    MaybeError Graph::AddBatchNorm(const op::BatchNorm* batchNorm) {
+        DAWN_TRY(CheckStatusCode(ANEURALNETWORKS_OP_FAILED, "nnapi Batch norm"));
+        return {};
+    }
+
+    MaybeError Graph::AddBinary(const op::Binary* binary) {
+        auto input0OpIndex = mGraphNodeMap[binary->Inputs()[0].Get()];
```

Comment:
You can use `primaryIndex `instead of `input0OpIndex`, `secondaryIndex `instead of `input1OpIndex`